TlmPacketizer Group Level Control Improvements (#4668)
nasa/fprime: Svc/TlmPacketizer/docs/sdd.md

Reviewer: Update this requirement to match the new approach, which isn't "all packets," it's the packets sent on the cycle they are specified to be sent on.

> The `Svc::TlmPacketizer` Component is used to store telemetry values written by other components. The values are stored in serialized form. TlmPacketizer differs from `Svc::TlmChan` in that it stores telemetry in defined packets instead of streaming the updates as they come. The defined packets are passed in as a table to the `setPacketList()` public method. When telemetry updates are passed to the component, they are placed at the offset in a packet buffer defined by the table. When the `run()` port is called, all the defined packets are sent to the output port with the most recent . This is meant to replace `Svc::TlmCham` for use cases where a more compact packet format is desired. The disadvantage is that all channels are pushed whether or not they have been updated.

Reviewer: Fix "most recent ." to "most recent values."

> Uses can change the individual rates at which groups per group instance are outputted upon a `run()` port call. Groups are configured with a MIN number of `run()` invocations to emit updated telemetry, and a MAX number of `run()` invocations that shall output a packet if it exceeds MAX. Packets are evaluated individually and have a counter since last invocation. MIN and MAX logic are selectable, for users that desire only "on change" telemetry, MAX invocations between telemetry points, both, or none at all.

Reviewer 2: I personally like the phrasing of "Sched ticks," "Rate Group ticks," or simply "ticks" to contrast other unscheduled, non-thead-providing port invocations

Reviewer 2: "each group on each output section can have its own telemetry rate resampling configuration"

Reviewer 2: "Packets can be sent on change with a rate limiting enforced minimum number of ticks between updates. Or packets can be sent with at a guaranteed rate of a maximum number of ticks between updates."

> Uses can change the individual rates at which groups per group instance are outputted upon a `run()` sched tick. Each group on each output section has independently configurable telemetry resampling rates. Packets can be sent on change with a rate limiting enforced minimum number of ticks between updates. Or packets can be sent with at a guaranteed rate of a maximum number of ticks between updates. Packets are evaluated individually and have a counter since last invocation. MIN and MAX logic are selectable, for users that desire only "on change" telemetry, MAX invocations between telemetry points, both, or none at all.

Reviewer 3: ```suggestion
Users can change the individual rates at which telemetry groups are output upon a `run()` sched tick for each telemetry section. Each group for each section has independently configurable telemetry resampling rates. Packets can be sent on change with a rate limiting enforced minimum number of ticks between updates. Alternatively, packets can be sent with at a guaranteed rate of a maximum number of ticks between updates. Packets are evaluated individually and have a counter since last invocation. MIN and MAX logic are selectable, for users that desire only "on change" telemetry, MAX invocations between telemetry points, both, or none at all.
```

> Uses can change the individual rates at which groups per group instance are outputted upon a `run()` sched tick. Each group on each output section has independently configurable telemetry resampling rates. Packets can be sent on change with a rate limiting enforced minimum number of ticks between updates. Or packets can be sent with at a guaranteed rate of a maximum number of ticks between updates. Packets are evaluated individually and have a counter since last invocation. MIN and MAX logic are selectable, for users that desire only "on change" telemetry, MAX invocations between telemetry points, both, or none at all.

Reviewer 3: "MIN and MAX logic are selectable, for users that desire only "on change" telemetry, MAX invocations between telemetry points, both, or none at all." Might need rephrasing.

#### 3.1.2 Ports

The `Svc::TlmPacketizer` component uses the following port types:

Port Data Type | Name | Direction | Kind | Usage
-------------- | ---- | --------- | ---- | -----
[`Svc::Sched`](../../Sched/docs/sdd.md) | Run | Input | Asynchronous | Execute a cycle to write changed telemetry channels
[`Fw::Tlm`](../../../Fw/Tlm/docs/sdd.md) | TlmRecv | Input | Synchronous Input | Update a telemetry channel
[`Fw::Com`](../../../Fw/Com/docs/sdd.md) | PktSend | Output | n/a | Write a set of packets with updated telemetry
`Svc::EnableSection` | controlIn | Input | Asynchronous | Enable / Disable sections of telemetry groups

#### 3.1.3 Terminology
Telemetry Point: An emitted value.

> Telemetry Channel: A tagged type and identifier for emitting telemetry points.

Reviewer 3: Also has time!

Telemetry Packets: A group of telemetry channels.

> Telemetry Group / Level: An identifier for a telemetry packet used to determine which packets get transmitted.

Reviewer 2: "A set of telemetry packets with shared control"

Author: Using implied "definition" from here: https://nasa.github.io/fpp/fpp-users-guide.html#Defining-Topologies_Telemetry-Packets_Telemetry-Packet-Sets

Reviewer: I would say "get transmitted together," since that's the intent of the telemetry groups. We should update the FPP user guide to match this.

> Telemetry Section: A group of telemetry groups.

Reviewer 2: "One resampling of telemetry groups. Each Section typically is destined for a different destination (e.g. realtime downlink vs store & forward)"

The implementation uses a hashing function to find the location of telemetry channels that is tuned in the configuration file `TlmPacketizerImplCfg.hpp`. See section 3.5 for description.

When a call to the `Run()` interface is called, the packet writes are locked and all the packets are copied to a second set of packets. Once the copy is complete, the packets writes are unlocked. The destination packet set gets updated with the current time tag and are sent out the `pktSend` port.

> Each telemetry group, depending on section and group configurations, are sent out on several indices of the `pktSend` port. As an example, a packet with group 1 with a TlmPacketizer configuration of 3 sections (0, 1, and 2), and 3 groups (0 and 3 inclusive), will be sent on 3 indices: 1, 5, and 9. Port invocations to `controlIn` or the command, `ENABLE_SECTION` are used to enable / disable each section, supporting downstream components that rely on different samplings of groups of telemetry. Each group instance is separately sampled from each other, allowing for individual rates per section and group.

Reviewer 2: I wouldn't call [0, 3] 3 groups. Either "max group number 3" or "4 groups"

Reviewer 2: "Since each section holds a continuous range of ports"

### 3.3 Scenarios

#### 3.3.1 External User Option

#### 3.3.2 Configuring Telemetry Group Rates Per Port

The `Svc::TlmPacketizer` is configurable to have multiple `PktSend` group outputs using the fpp constant, `MAX_CONFIGURABLE_TLMPACKETIZER_GROUP`. Doing so allows each packet group have different configurations for each `PktSend` output section.

> `PktSend` output is ordered by [section][group], where within each section telemetry groups (inclusive) are emitted. Each group within each section are separately sampled, and have their own configuration rates and logic independent from each other. 

Reviewer 2: ```suggestion
`PktSend` output is ordered by `[section][group]`, where within each section telemetry groups (inclusive) are emitted. Each group within each section are separately sampled, and have their own configuration rates and logic independent from each other.
```

Reviewer 2: I don't think you need the word `inclusive` here

> Each group is configured with min/max delta parameters, as well as a logic gate determining its output rate behavior. Min Delta is the least number of `Run()` invocations between successive packets before an updated packet is allowed to be sent. Updated packets will not be sent until their counter reaches Min Delta. This mitigates against telemetry spam and allows users to configure faster updated and sent telemetry compared to updates that don't require to be as frequently updated. Max Delta is the maximum number of `Run()` invocations between successive packets. Upon reaching this counter, the packet would be sent, regardless of change.

Reviewer 2: Make sure you define that delta is a number of ticks/run invocations the first time you use the term

Reviewer 2: I would probably specify that an "updated packet" is a "packet with an updated channel" the first time that term is used

Reviewer 2: I would potentially be more specific on the min delta side that "This mitigates telemetry spam while allowing users to benefit from asynchronously updating channels (those that don't occupy a set schedule)"

Each telemetry group per section is configured with a `RateLogic` parameter:
* `SILENCED`: Packet will never be sent
* `EVERY_MAX`: Packet will be sent every Max Delta intervals

> * `ON_CHANGE_MIN`: Packet will only be sent on updates at at least Min Delta intervals.

Reviewer 3: No faster than min delta intervals, yes?

Author: In the current implementation, MAX will take priority over MIN if MIN is greater than MAX (when set to `ON_CHANGE_MIN_OR_EVERY_MAX`). Will add to sdd.

* `ON_CHANGE_MIN_OR_EVERY_MAX`: Packet will be sent on updates at at least Min Delta intervals, or at most Max Delta intervals.

Groups are configured using the `SET_GROUP_DELTAS` command.

#### 3.3.3 Switching Telemetry Sections

> `Svc::TlmPacketizer` also has a configurable constant `NUM_CONFIGURABLE_TLMPACKETIZER_SECTIONS` that allows separately streamed outputs of packets of the same group. This is useful for downstream components that would handle different telemetry groups at different rates. Examples of this configuration includes live telemetry throughput, detailed sim reconstruction, or critical data operations. 

Reviewer 2: ```suggestion
`Svc::TlmPacketizer` also has a configurable constant `NUM_CONFIGURABLE_TLMPACKETIZER_SECTIONS` that allows separately streamed outputs of packets of the same group. This is useful for downstream components that would handle different telemetry groups at different rates. Examples of this configuration includes live telemetry throughput, detailed sim reconstruction, or critical data operations.
```

#### 3.1.2 Ports

The `Svc::TlmPacketizer` component uses the following port types:

Port Data Type | Name | Direction | Kind | Usage
-------------- | ---- | --------- | ---- | -----
[`Svc::Sched`](../../Sched/docs/sdd.md) | Run | Input | Asynchronous | Execute a cycle to write changed telemetry channels
[`Fw::Tlm`](../../../Fw/Tlm/docs/sdd.md) | TlmRecv | Input | Synchronous Input | Update a telemetry channel
[`Fw::Com`](../../../Fw/Com/docs/sdd.md) | PktSend | Output | n/a | Write a set of packets with updated telemetry
`Svc::EnableSection` | controlIn | Input | Asynchronous | Enable / Disable sections of telemetry groups

> #### 3.1.3 Terminology

Reviewer: This text is getting rendered as a run-on sentence. Put a carriage return between lines or put in a table to get a definition per row.

Telemetry Point: An emitted value.
Telemetry Channel: A tagged type, identifier, and timestamp for emitting telemetry points.
Telemetry Packets: A group of telemetry channels.

> Telemetry Group / Level: An identifier for a telemetry packet used to determine which packets get transmitted.

Reviewer: It should be "An identifier that the specifies a set of packets to be transmitted togther."

The implementation uses a hashing function to find the location of telemetry channels that is tuned in the configuration file `TlmPacketizerImplCfg.hpp`. See section 3.5 for description.

When a call to the `Run()` interface is called, the packet writes are locked and all the packets are copied to a second set of packets. Once the copy is complete, the packets writes are unlocked. The destination packet set gets updated with the current time tag and are sent out the `pktSend` port.

> Each telemetry group, depending on section and group configurations, are sent out on several indices of the `pktSend` port. Since each section holds a continuous range of ports, a packet with group 1 with a TlmPacketizer configuration of 3 sections (0, 1, and 2), and a max group of 3 (0 and 3 inclusive), will be sent on 3 indices: 1, 5, and 9. Port invocations to `controlIn` or the command, `ENABLE_SECTION` are used to enable / disable each section, supporting downstream components that rely on different samplings of groups of telemetry. Each group instance is separately sampled from each other, allowing for individual rates per section and group.

Reviewer: Reading from the beginning, I have no idea why 0,1,2 would go to ports 1, 5, 9.  Is this an example? Is there an algorithmic mapping? Should this go in the algorithms section?

### 3.3 Scenarios

#### 3.3.1 External User Option

#### 3.3.2 Configuring Telemetry Group Rates Per Port

> 

Reviewer: How does one configure this? Is it a data structure? If so, show the data structure and its specification prior to describing what goes in it?

The `Svc::TlmPacketizer` is configurable to have multiple `PktSend` group outputs using the fpp constant, `MAX_CONFIGURABLE_TLMPACKETIZER_GROUP`. Doing so allows each packet group have different configurations for each `PktSend` output section.

`PktSend` output is ordered by `[section][group]`, where within each section telemetry groups are emitted. Each group within each section are separately sampled, and have their own configuration rates and logic independent from each other.

> Each group is configured with min/max delta parameters, as well as a logic gate determining its output rate behavior. Deltas are counters of sched ticks invoked through the `run()` port. Min Delta is the least number of `Run()` invocations between successive packets before a packet with an updated channel is allowed to be sent. Updated packets will not be sent until their counter reaches Min Delta. This mitigates against telemetry spam while allowing users to benefit from asynchronously updating channels (e.g. those that don't occupy a set schedule). Max Delta is the maximum number of `Run()` invocations between successive packets. Upon reaching this counter, the packet would be sent, regardless of change.

Reviewer: Nitpick: Spell out "sched"

Each telemetry group per section is configured with a `RateLogic` parameter:
* `SILENCED`: Packet will never be sent
* `EVERY_MAX`: Packet will be sent every Max Delta intervals
* `ON_CHANGE_MIN`: Packet will only be sent on updates at at least Min Delta intervals.
* `ON_CHANGE_MIN_OR_EVERY_MAX`: Packet will be sent on updates at at least Min Delta intervals, or at most Max Delta intervals.

Note: `MAX` will take priority over `MIN` if less than `MIN` and the telemetry group is set to `ON_CHANGE_MIN_OR_EVERY_MAX`. 

> Groups are configured using the `CONFIGURE_GROUP_RATES` command.

Reviewer: Are they configured away from a default, or do you have to issue commands to get any telemetry at all?

Author: Telemetry packets are by default, configured with MIN = 0 and ON_CHANGE_MIN. This preserves the behavior of the original TlmPacketizer behavior, and FSW will have telemetry even when unconfigured.

For advanced behavior, such as setting MIN and MAX attributes to each tlm group, A command must be sent to configure it so.
